pkg/skaffold/build/cache/once.go

```diff
@@ -0,0 +1,64 @@
+/*
+Copyright 2020 The Skaffold Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package cache
+
+import (
+	"sync"
+
+	"github.com/sirupsen/logrus"
+
+	"github.com/GoogleContainerTools/skaffold/pkg/skaffold/schema/latest"
+)
+
+// hashOnce ensures that each artifact hash is computed at most once.
+// This is useful when several artifacts request the hash for a common required artifact.
+type hashOnce interface {
+	get(a *latest.Artifact, hashFunc func(*latest.Artifact) (string, error)) (string, error)
+}
```

Comment:
Is the single-hash really part of the interface?  Or is it an implementation detail?  Do callers need to get the hash of an artifact with multiple hash-functions?  Could this be an `artifactHasher` that is instantiated with the hash-function?

Reply:
The `artifactHasher` is already embedded in `cache` as the function `getHashForArtifact`. I don't want to refactor too much of the code so I think it makes sense for `hashOnce` as a wrapper. 

It doesn't have to have an interface tied to hash really and I can move it to `utils` package and change it to:
```go
type Once {}
// Do calls the function f iff it's being called the first time for a specific key (evaluated from the key function).
// If it's called multiple times for the same key only the first call will execute and store the result of f.
// All other calls will be blocked until the running instance of f returns and all of them receive the same result.
func(*Once) Do(key func() interface{}, f func() interface{}) interface{}
```

WDYT?

Reply:
discussed over DM.